UOP1_Project/Assets/Scripts/Systems/SettingsSystem.cs

```diff
@@ -0,0 +1,278 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using TMPro;
+using UnityEngine;
+using UnityEngine.Rendering;
+using UnityEngine.UI;
+
+public class SettingsSystem : MonoBehaviour
```

Comment:
I'm not sure that having a single class be responsible for all the different kinds of settings is a great idea... there's a lot of different things in here, and if we add more settings over time this class will become a bit of a dumping-ground.

Have you considered making separate components for each of the different kinds of control - e.g. a ResolutionsDropdownController component, which you attach to the resolutions dropdown GameObject, a LanguageDropdownController component that you attach to the languages dropdown GameObject, etc?

You could still have the SettingsSystem provide a facade for accessing all the settings if you want, but the logic around how to populate each setting and how to apply it would be isolated into separate classes that would be easier to maintain.

Reply:
In my experience, separating a logical block as "Settings Systems" into too small pieces is making it too complex to maintain. Especially when adding more settings later on. Each having its own class with maybe 1-3 lines doesn't make it worth it for me to create dozens of classes, instead of letting one settings class do the job.

Also, components like "ResolutionsDropdownController" would only redirect a method call to the Main settings class, which you can skip and directly call in on said settings class.

But splitting it up into Sub classes like "GeneralSettingsComponent", "GraphicsSettingsComponent", "AudioSettingsComponent" etc would be an option. Currently these logical blocks are separated by a #region statement in the SettingsSystem class.

Reply:
> Also, components like "ResolutionsDropdownController" would only redirect a method call to the Main settings class, which you can skip and directly call in on said settings class.

How so?

Reply:
Assuming that the settings are not saved right away, but fist saved when hitting the save button (advanced graphics settings), you would have to temporarily save the settings in some sort of variable.

So of course there needs to be a central class holding this information and that would be the Main Settings class. That's why there has to be a redirection.

Without redirection, each component has to handle the save/cancel operation itself, which leads to the same code in dozens of small components. Also loading previously saved data would have had the same problem. All in all there are several more reasons which add up to a more maintainable "One Class"-Solution.

Reply:
when you referred to storing data when the button isn't clicked maybe we can get the data straight from the object?

We will obviously be changing the properties of the object to give visual response but we can take the data straight from the Object it self that has been changed but getting the text or anything like that instead of making variables?

Reply:
> Assuming that the settings are not saved right away, but fist saved when hitting the save button (advanced graphics settings), you would have to temporarily save the settings in some sort of variable.

Yes...

> So of course there needs to be a central class holding this information

...no.

Each separate controller would hold that state itself. Then you'd have e.g. an Apply() method on each controller, that the overall controller for the page would call when you hit the 'Apply' button. It's not the same code in each controller because the way of applying each setting is different - one controller is calling SetQualityLevel, another is doing `vsyncCount = 1;`, etc.

Reply:
But is it worth creating a whole class just for the sake of calling QualitySetting.antiAliasing = x? In my opinion, it's not, but this is a taste issue rather than an implementation one. Both would work equally.

Also, this "Apply" button would have to call the "apply" in a center class, which then calls it on all sub-components or you would have to add each apply call directly to the button click event itself (which honestly I'm not a big fan of because the button events are much more prone to errors when developing).

Reply:
> But is it worth creating a whole class just for the sake of calling QualitySetting.antiAliasing = x?

Antialiasing might be that simple, but I think that in some other cases - such as resolutions - it's a bit more involved: the class would not just apply the resolution, it would also be responsible for populating the dropdown with the list of supported resolutions, identifying which resolution is the current one, converting each Resolution object to a good-looking string, etc. In the case of antialiasing, I agree there's very little code to write if we are just setting QualitySettings, but it might be more involved if we later want to use the antialiasing features in the postprocessing stack, instead of the 'built-in' antialiasing.

Note that each individual setting control would not necessarily need to have exactly one control associated with it - if we wanted to offer a dropdown for 'which AA technique' as well as a slider for 'what level' then it would probably make sense for those to be driven by the same controller component, especially if e.g. the slider is enabled/disabled based on the dropdown. For something like the input bindings it's plausible that all of them would be associated with the same controller.

> Also, this "Apply" button would have to call the "apply" in a center class, which then calls it on all sub-components or you would have to add each apply call directly to the button click event itself (which honestly I'm not a big fan of because the button events are much more prone to errors when developing).

Yes, what I imagine is that you'd have an abstract base class `SettingController` that offers the Apply() method, and each individual controller class would derive from it, implementing that method. Then the overall controller for the page could just do `GetComponentsInChildren<SettingController>()` to get them all, then loop through and call `Apply()` on each one.

Reply:
What i see here is what our company did and learned from in the past. Overly abstracting things in the very beginning and adding more work for a feature that otherwise would have a very simple way of implementation. I don't say that it would come in handy in certain situations, but at this point we cant say for what this is going to be used. And so we should stick to "keep it simple" and save us unnececary work.

Also calling any "GetComponent" code, even if its just UI is not a good idea in general. I always try to avoid using this when I can. For example it wont work ob objects which are deactivated (which happens quite often in UI).

Reply:
>  I don't say that it would come in handy in certain situations, but at this point, we can't say for what exactly this is going to be used.

21 years of experience tells me that we're going to need it - but sure, it would be good if we had something to go on more than gut feelings. Maybe we should have done some actual design on this first 😂 

Anyway, it's your PR, I'm just here to try and help everyone learn and explore possibilities. If it's not a direction you want to go in, so be it. :)

> GetComponentInChildren doesn't work on inactive objects

It does if you pass `true` for the `includeInactive` parameter.

Reply:
This is indeed a great parameter. This should be possible in Transform.Find() as well! (unfortunately it's not 😂 )

I've just created a new PR (https://github.com/UnityTechnologies/open-project-1/pull/67) which in some way utilizes what you have said already. I split up the logic in sub-components (General, Graphics, Audio, Input) which do the job of populating the dropdowns, etc.
It's not exactly how you described it, but I felt like this is more like I would implement it. Maybe we can talk about the pros and cons of this. Btw I didn't want to be rude in any way. I appreciate your contribution and feedback to my code :)